### PR TITLE
New definitions for nat-upnp

### PR DESCRIPTION
**Pull request by SimplyLinn**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [X] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/nat-upnp/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+/**
+ * Main client interface.
+ */
+export interface Client {
+    /**
+     * Create a new port mapping
+     * @param options Options for the new port mapping
+     * @param callback Callback to be run when completed, or on error
```
Comment (peterblazejewicz) [inline types/nat-upnp/index.d.ts:161]:
```suggestion
     * @param [callback] Callback to be run when completed, or on error
```

File: types/nat-upnp/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+     */
+    portMapping(options: NewPortMappingOpts, callback: CB<RawResponse>): void;
```
Comment (peterblazejewicz) [inline types/nat-upnp/index.d.ts:163]:
```suggestion
    portMapping(options: NewPortMappingOpts, callback?: CB<RawResponse>): void;
```
https://github.com/indutny/node-nat-upnp/blob/9c80722df962ce1984946038e702e8eaccd0e97a/lib/nat-upnp/client.js#L31

File: types/nat-upnp/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    /**
+     * Remove a port mapping
+     * @param options Specify which port mapping to remove
+     * @param callback Callback to be run when completed, or on error
```
Comment (peterblazejewicz) [inline types/nat-upnp/index.d.ts:167]:
```suggestion
     * @param [callback] Callback to be run when completed, or on error
```

File: types/nat-upnp/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+     */
+    portUnmapping(options: DeletePortMappingOpts, callback: CB<RawResponse>): void;
```
Comment (peterblazejewicz) [inline types/nat-upnp/index.d.ts:169]:
```suggestion
    portUnmapping(options: DeletePortMappingOpts, callback?: CB<RawResponse>): void;
```
https://github.com/indutny/node-nat-upnp/blob/9c80722df962ce1984946038e702e8eaccd0e97a/lib/nat-upnp/client.js#L56

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
Those callbacks seem to be optional, otherwise sample code from README won't compile:
```ts
client.portUnmapping({
  public: 12345
});
```
Otherwise looks good

**Comment by SimplyLinn** [comment]
Oh, must've stared myself blind on the implementation when I was typing, totally missed the null-checks in those. Good catch!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@SimplyLinn thx for contribution!

**Pull request merged**